pkg/inframetadata/internal/hostmap/hostmap.go

```diff
@@ -0,0 +1,89 @@
+package hostmap
+
+import (
+	"fmt"
+	"sync"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.uber.org/multierr"
+
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/gohai"
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/payload"
+)
+
+// HostMap stores a map from hostnames to host metadata payloads.
+// Host metadata payloads can be updated with the information from a given resource.
+// At any time, the internal map can be extracted and the map will be cleared out.
+type HostMap struct {
+	// mutex for the host map.
+	mutex sync.Mutex
+	// hosts map
+	hosts map[string]payload.HostMetadata
+}
+
+// New creates a new HostMap
+func New() (*HostMap, error) {
+	return &HostMap{
+		mutex: sync.Mutex{},
+		hosts: make(map[string]payload.HostMetadata),
+	}, nil
+}
+
+// Update the information about a given host by providing a resource.
+// The function reports:
+//   - Whether the information about the `host` has changed
+//   - Any non-fatal errors that may have occurred during the update
+//
+// Partial modifications will still be applied even with non-fatal errors.
```

Comment:
Imagine you get a payload with the following fields:

- `host.name` with a string value `myhostname`
- `os.kernel` with a string value `Linux`
- `host.arch` with a string value `x86_64`
- `os.description` with a boolean value of `true` (this is wrong, should be a string)

What should we do? I see two options:

1. Update the fields that have the correct values (i.e. all of them but `os.description`), and return an error signaling that the type of `os.description` is wrong
2. Don't update any fields, since one of them is wrong, and return an error signaling that the type of `os.description` is wrong

Option (1) is what we are currently doing (the `os.description` type mismatch is non-fatal), option (2) would be what we do if we restore the original value.

I feel like option (1) is better in that the error is local to `os.description` and so the user experience is only partially degraded (one field missing vs. all fields missing). Do you think we should do option (2)? Or maybe the comment was unclear?

Reply:
Thanks, I think the current behavior makes more sense. Maybe update the comment to also mention order doesn't matter and all fields without errors will be applied. I.e. in your example, if it's `os.description`, `host.name`, `host.arc`, `os.kernel` it will yield the same result.

Reply:
Updated comment to clarify: https://github.com/DataDog/opentelemetry-mapping-go/blob/aec0871cf47d1dac76eebce1378b034ddce47c99/pkg/inframetadata/internal/hostmap/hostmap.go#L94-L100